File: concepts/01_values_and_comparison/01_truthiness/nodejs/test_truthiness.mjs
// 横向概念 001｜真假值与逻辑运算结果。
// 共同问题：数值零、空文本、空集合和空值如何判断；自定义对象能否定义真假；
// 逻辑运算符返回布尔值还是原操作数。
//
// polyglot-concept: truthiness
// polyglot-related: languages/nodejs/language/test_001_primitive_values_numeric_models_and_equality.mjs

import assert from 'node:assert/strict';
import test from 'node:test';

test('零、空文本和 nullish 值为假', () => {
  const falsyValues = [0, -0, 0n, '', null, undefined, Number.NaN];

  assert.ok(falsyValues.every((value) => Boolean(value) === false));
  assert.equal(Boolean(1), true);
  assert.equal(Boolean('0'), true);
});

test('空数组、空对象和空集合仍然为真', () => {
  assert.equal(Boolean([]), true);
  assert.equal(Boolean({}), true);
  assert.equal(Boolean(new Set()), true);
  assert.equal(Boolean(new Map()), true);

  // ToBoolean 只看“对象”这一类型，不询问 length、size 或集合内容。
});

test('对象不能用转换钩子覆盖 ToBoolean', () => {
  let conversionCalls = 0;
  const value = {
    [Symbol.toPrimitive]() {
      conversionCalls += 1;
      return 0;
    },
  };

  assert.equal(Boolean(value), true);
  assert.equal(conversionCalls, 0);
});

test('逻辑运算符返回被选中的原操作数', () => {
  const emptyArray = [];
  const fallback = { source: 'default' };
  const payload = ['ready'];

  assert.equal(emptyArray || fallback, emptyArray);
  assert.equal('' || fallback, fallback);
  assert.equal(payload && fallback, fallback);
  assert.deepEqual(0 && fallback, 0);

  // 第一条故意证明空数组本身已是真值；它不会像 Python 的 [] 那样选择 fallback。
});

File: concepts/001_truthiness/python/test_truthiness.py
"""横向概念 001｜真假值与逻辑运算结果。

共同问题：数值零、空文本、空集合和空值如何判断；自定义对象能否定义真假；
逻辑运算符返回布尔值还是原操作数。
"""

# polyglot-concept: truthiness
# polyglot-related: languages/python/language/test_001_truth_value_testing.py


class ExplicitFlag:
    def __init__(self, enabled):
        self.enabled = enabled

    def __bool__(self):
        return self.enabled


class SizedQueue:
    def __init__(self, items):
        self.items = list(items)

    def __len__(self):
        return len(self.items)


def test_zero_empty_text_and_none_are_false():
    assert bool(0) is False
    assert bool(1) is True
    assert bool("") is False
    assert bool("0") is True
    assert bool(None) is False


def test_empty_builtin_collections_are_false():
    assert bool([]) is False
    assert bool({}) is False
    assert bool(set()) is False
    assert bool([0]) is True


def test_custom_objects_can_define_truth_with_two_protocol_levels():
    assert bool(object()) is True
    assert bool(ExplicitFlag(False)) is False
    assert bool(ExplicitFlag(True)) is True
    assert bool(SizedQueue([])) is False
    assert bool(SizedQueue(["job"])) is True


def test_and_or_return_the_selected_operand():
    fallback = {"source": "default"}
    payload = ["ready"]

    assert ([] or fallback) is fallback
    assert (payload and fallback) is fallback
    assert ([] and fallback) == []
